steel/src/contract.rs

```diff
@@ -0,0 +1,441 @@
+#[cfg(feature = "host")]
+use crate::host::{db::ProofDb, provider::Provider, HostViewCallEnv};
+use crate::{EvmHeader, GuestViewCallEnv};
+use alloy_primitives::{keccak256, Address, B256, U256};
+
+use alloy_sol_types::{SolCall, SolType};
+
+use revm::{
+    primitives::{
+        AccountInfo, Bytecode, ExecutionResult, HashMap, ResultAndState, SuccessReason, TransactTo,
+    },
+    Database, Evm,
+};
+use std::{convert::Infallible, fmt::Debug, marker::PhantomData, mem, rc::Rc};
+
+/// A type pointing at a contract using the environment and contract address initialized with. This
+/// contract is not typesafe, be sure the contract deployed at the address matches the ABI you use
+/// to make calls.
+///
+/// To preflight calls in the host to build the proof, use [Contract::preflight], using the env
+/// from [EthViewCallEnv::from_rpc] or [ViewCallEnv::new].
+///
+/// To initialize in the guest, use [Contract::new], with the environment constructed through
+/// [ViewCallInput::into_env].
+///
+/// # Examples
+///
+/// ```no_run
+/// use risc0_steel::{ethereum::EthViewCallEnv, Contract};
+/// use alloy_primitives::{address};
+/// use alloy_sol_types::sol;
+///
+/// # fn main() -> anyhow::Result<()> {
+/// let contract_address = address!("dAC17F958D2ee523a2206206994597C13D831ec7");
+/// sol! {
+/// #[derive(Debug, PartialEq, Eq)]
+/// interface IERC20 {
+///    function balanceOf(address account) external view returns (uint);
+/// }
+/// }
+///
+/// let get_balance = IERC20::balanceOfCall {
+///     account: address!("F977814e90dA44bFA03b6295A0616a897441aceC"),
+/// };
+///
+/// // Host:
+///
+/// let mut env = EthViewCallEnv::from_rpc("https://ethereum-rpc.publicnode.com", None)?;
+/// let mut contract = Contract::preflight(contract_address, &mut env);
+/// contract.call_builder(&get_balance).call()?;
+///
+/// let view_call_input = env.into_zkvm_input()?;
+///
+/// // Guest
+///
+/// let view_call_env = view_call_input.into_env();
+/// let contract = Contract::new(contract_address, &view_call_env);
+/// contract.call_builder(&get_balance).call();
+///
+/// # Ok(())
+/// # }
+/// ```
+///
+/// [EthViewCallInput::into_env]: ethereum::EthViewCallInput::into_env
+/// [EthViewCallEnv::from_rpc]: ethereum::EthViewCallEnv::from_rpc
+pub struct Contract<E> {
+    address: Address,
+    env: E,
+}
+
+impl<'a, H> Contract<&'a GuestViewCallEnv<H>> {
+    pub fn new(address: Address, env: &'a GuestViewCallEnv<H>) -> Self {
+        Self { address, env }
+    }
+
+    /// Initialize a call builder to execute a call on the contract. For more information on usage,
+    /// see [Contract].
+    pub fn call_builder<C: SolCall>(&self, call: &C) -> ViewCallBuilder<&GuestViewCallEnv<H>, C> {
+        ViewCallBuilder::new_sol(self.env, self.address, call)
+    }
+}
+
+#[cfg(feature = "host")]
```

Comment:
So far I have tried to have everything that depends on the host feature in the `host` module, this also eliminates the need for those ugly configs for imports.
But of course, this approach would reduce code locality, and that is a tradeoff. So let's see how everything looks after the consolidations and see which one makes more sense.

Reply:
Agree, I'm right on the fence with this, and just made a judgement call of what _seemed_ most clean. No opinions here, so happy to switch if anyone shares an opinion!